nx/lib/nx/defn/expr.ex

```diff
@@ -816,6 +816,12 @@ defmodule Nx.Defn.Expr do
     end
   end
 
+  def optional(expr, name, [out, %{data: %{context: context}} | _] = args) do
+    out = %T{names: [], shape: {}, type: out.type}
```

Comment:
The names and shapes and types here are wrong. It should be the same as the output.

Reply:
I think this function should be something like this:

```elixir
# args is already without out
def optional(%{data: %{context: context}} = expr, name, args) do
  call = expr(expr, context, name, args)
  expr(expr, context, :default_implementation, [call, expr])
end
```

Then the grad will happen on the second argument. :)

You also want to change the inspect implementation to only consider the first argument. You don't want to show the `default_implementation` nor the second argument.

Reply:
I see! This makes sense